google/oauth2/credentials.py

```diff
@@ -291,34 +291,36 @@ def with_quota_project(self, quota_project_id):
             scopes=self.scopes,
             default_scopes=self.default_scopes,
             granted_scopes=self.granted_scopes,
-            quota_project_id=quota_project_id,
+            quota_project_id=self.quota_project_id,
             rapt_token=self.rapt_token,
             enable_reauth_refresh=self._enable_reauth_refresh,
             trust_boundary=self._trust_boundary,
             universe_domain=self._universe_domain,
             account=self._account,
         )
+        cred._cred_file_path = self._cred_file_path
+        return cred
+
+    @_helpers.copy_docstring(credentials.Credentials)
+    def get_cred_info(self):
+        if self._cred_file_path:
```

Comment:
This requires the auth lib getting the id token first then decoding the id token. This may cause implementation difficulties for other languages due to the complexity, so I didn't include it in the approved language neutral doc. This could be a future add-on feature though.

Reply:
email field is present in the access token right? We need to decrypt the AT by calling the token info endpoint.

Reply:
Based on my testing before, I think the token info endpoint only works if the access token is still valid, otherwise it will return an error response saying token is invalid. ID token is the most reliable way (this is what gcloud does).

Reply:
We could call the token info endpoint with the access token as soon as we get an access token for the very first time and store it in the credential?

Reply:
Done.

Added a `_set_account_from_access_token` method, if self._account doesn't exist, then it calls token info endpoint to get the email, then set self._account to the email address. This way we only need to talk to token info endpoint once during the cred's lifetime. This method is called in the `refresh` method after refresh.

When adding cred_info, we look at if self._account exists, if so then we add it to cred info json as the principal.